File: frontend/src/app/services/topic.service.ts
import { Injectable } from '@angular/core';
import { Topic } from '../dtos/topic';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { Globals } from '../global/globals';

@Injectable({
    providedIn: 'root'
})
export class TopicService {

    private topicBaseUri: string = this.globals.backendUri + '/topics';

    constructor(private httpClient: HttpClient, private globals: Globals) {
    }

    /**
     * Loads all topics from the backend
     */
    getTopics(): Observable<Topic[]> {
        return this.httpClient.get<Topic[]>(this.topicBaseUri);
    }

    /**
     * Get a container from the backend by id
     *
     * @param id of the container
     */
    getByID(id: number): Observable<Topic> {
        console.log('Get a container by the id ' + id);
        return this.httpClient.get<Topic>(this.topicBaseUri + '/' + id);
    }

    /**
     * Update a container in the backend
     *
     * @param topic to update
     */
    update(topic: Topic): Observable<Topic> {
        console.log('Updated container with id ' + topic.id);
        return this.httpClient.put<Topic>(this.topicBaseUri + '/' + topic.id, topic);
    }

    /**
     * Create a container in the backend
     *
     * @param topic to create
     */
    create(topic: Topic): Observable<Topic> {
        console.log('Create container with name ' + topic.name);
        return this.httpClient.post<Topic>(this.topicBaseUri, topic);
    }

    /**
     * Deletes a container in the backend
     *
     * @param id of topic to delete
     */
    deleteTopic(id: number): Observable<Topic> {
        console.log('Delete container with id ' + id);
        return this.httpClient.delete<Topic>(this.topicBaseUri + '/' + id);
    }


}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/TopicService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.ContainerTopic;
import at.ac.tuwien.sepm.groupphase.backend.exception.ConflictException;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ValidationException;

import java.util.List;

public interface TopicService {

    /**
     * Get all topics.
     *
     * @return all topics
     */
    List<ContainerTopic> getAll();

    /**
     * Get a topic by id.
     *
     * @param id of the topic
     * @return the topic
     */
    ContainerTopic getById(Long id);

    /**
     * Create a new topic.
     *
     * @param containerTopic to create
     * @return the created containerTopic
     */
    ContainerTopic create(ContainerTopic containerTopic) throws ValidationException, ConflictException;

    /**
     * Update an existing topic.
     *
     * @param id of the topic to update
     * @param containerTopic to update
     * @return updated containerTopic
     */
    ContainerTopic update(Long id, ContainerTopic containerTopic) throws ValidationException, ConflictException;

    /**
     * Delete an existing topic by its ID.
     *
     * @param id of the topic to delete
     * @throws NotFoundException if topic is not found
     */
    void deleteTopic(Long id);
}
